ui: add new cluster explorer page to advanced debug

jasonlmfong opened the pull request:
This new page contains a new hexagon heatmap powered by d3.
Each hexagon represents a single node in the cluster, the heatmap is colored by the user-selected
metric: cpu percentage or write bytes.

When a node hexagon is selected, a uplot histogram is shown.
This histogram offers further drill-down of the same metric on ranges on the selected node.
This histogram component is similar to the existing BarGraphTimeSeries component.

Epic: None
Release note: None

dhartunian reviewed (changes requested):
Why is there no CSS in this diff? All the styling is inline which is fine for an experiment, but it can introduce some challenges with integrating into our visual style.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-:-OgXh7Y9BH8lOF7MqAIx:bm1geq3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @jasonlmfong, and @kyle-a-wong)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXgrtk-taTUF7HQped:-OgXgrtk-taTUF7HQpee:b-bqwevj):*
I'd appreciate a little rundown here of how this code is structured and what it does. There's nothing either here or in the component files that explains what the "Cluster Explorer" is. It's a very generic name.
___
*[`pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXg6m285TnGZeUQB4q:-OgXg6m285TnGZeUQB4r:b-biunr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js#L9)):*
> ```JavaScript
> // Copyright 2021, Observable Inc.
> // Released under the ISC license.
> // https://observablehq.com/@d3/color-legend
> ```

Please follow documented process for including open source code in our repo: https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/4246536270/How+to+copy+open-source+code+into+the+cockroach+repository

We can't use npm for this dependency?
___
*[`package-lock.json` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXhQVB4ZUl4niaqvva:-OgXhQVB4ZUl4niaqvvb:bpp4djq) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/package-lock.json#L46)):*
> ```JSON
>     }
>   }
> }
> ```

this file shouldn't exist.
___
*[`pkg/ui/pnpm-lock.yaml` line 20901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXhLI27hCKnuapLUa0:-OgXhLI27hCKnuapLUa1:bgp1ime) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/pnpm-lock.yaml#L20901)):*
> ```YAML
>       prop-types: ^15.6.0
>       react: 16.12.0
>       react-dom: 16.12.0
> ```

why did the package lock change? do we need this diff?
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/debug/index.tsx` line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXfit2DQyFeLNHG6WC:-OgXfit2DQyFeLNHG6WD:bfhnn06) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/workspaces/db-console/src/views/reports/containers/debug/index.tsx#L310)):*
> ```XML
>           url="#/cluster-explorer"
>           note="Explore and analyze your cluster with interactive node and range visualization."
>         />
> ```

I don't think this page is at a level of polish that it should get a prominent link at the top. This is an experiment that we want to keep and iterate on, but anything at the top here requires documentation because it's really being presented as a thing users should actively engage with.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
made a styl file

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-:-OgYk-ho70wpFPsDd2SG:bhr3p0j)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXgrtk-taTUF7HQped:-OgYHxFICZqE_XgL09TJ:b2nhysk):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'd appreciate a little rundown here of how this code is structured and what it does. There's nothing either here or in the component files that explains what the "Cluster Explorer" is. It's a very generic name.
</blockquote></details>

ack
___
*[`package-lock.json` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXhQVB4ZUl4niaqvva:-OgYHv85-cZqvecFvvrY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/package-lock.json#L46)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

this file shouldn't exist.
</blockquote></details>

done
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/debug/index.tsx` line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXfit2DQyFeLNHG6WC:-OgYH4518i2ABUE4GsyZ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/workspaces/db-console/src/views/reports/containers/debug/index.tsx#L310)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't think this page is at a level of polish that it should get a prominent link at the top. This is an experiment that we want to keep and iterate on, but anything at the top here requires documentation because it's really being presented as a thing users should actively engage with.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`pkg/ui/pnpm-lock.yaml` line 20901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXhLI27hCKnuapLUa0:-OgYrGJb5tDY6xrAtswE:b22zcqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/pnpm-lock.yaml#L20901)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why did the package lock change? do we need this diff?
</blockquote></details>

this is dependent on the below change with legend.js
___
*[`pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXg6m285TnGZeUQB4q:-OgYrF9s881Clz-lOTN5:bh316yx) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js#L9)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Please follow documented process for including open source code in our repo: https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/4246536270/How+to+copy+open-source+code+into+the+cockroach+repository

We can't use npm for this dependency?
</blockquote></details>

I played around with this a bit, looks like the npm equivalent is not available in our mirror, so we end up having to include this code ourselves
the alternative would be to try to include that in our mirror of npm deps

also something about the go philosophy of a little copying


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`pkg/ui/pnpm-lock.yaml` line 20901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXhLI27hCKnuapLUa0:-Ogbl_m344ASRtlGUn4M:bx2ug49) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/pnpm-lock.yaml#L20901)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

this is dependent on the below change with legend.js
</blockquote></details>

the interpolate package was needed to add the legend
___
*[`pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXg6m285TnGZeUQB4q:-Ogblobc8rlouKQRlWEU:b-c1idz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ec64aac8a369420548892a96c10dd602b9de71/pkg/ui/workspaces/db-console/src/views/explorer/components/legend.js#L9)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

I played around with this a bit, looks like the npm equivalent is not available in our mirror, so we end up having to include this code ourselves
the alternative would be to try to include that in our mirror of npm deps

also something about the go philosophy of a little copying
</blockquote></details>

cleaned it up and we don't need this anymore, extracted the minimal amount of legend code


<!-- Sent from Reviewable.io -->

dhartunian reviewed (changes requested):
> @dhartunian reviewed 5 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @jasonlmfong, and @kyle-a-wong)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXgrtk-taTUF7HQped:-OgbpRJH3TttPlzxRcZw:bfy9hmf):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

ack
</blockquote></details>

Can you add a bit more detail? I think it should explain:

1. There's a heatmap where each node is represented by a hexagon.
2. User has ability to select a metric which determines what the colors on the hexagons display
3. Details about what metric is in the histogram. Same as the one in the hexagons?

It's not entirely clear how the selected metric is shown across nodes and also in the histogram across ranges. An explanation of your intent here would be handy.
___
*[`-- commits` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-Ogbq1ks1IVJpL9wDd_y:-Ogbq1ks1IVJpL9wDd_z:bjl9a7n):*
This page should not be documented in a release note. It's not ready for customer consumption.
___
*[`pkg/ui/workspaces/db-console/src/views/explorer/explorer.module.styl` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgbpGM18DlbUXHnP4dj:-OgbpGM18DlbUXHnP4dk:bgluebg) ([raw file](https://github.com/cockroachdb/cockroach/blob/93de803347c22b056f79653112b9340d02197ca3/pkg/ui/workspaces/db-console/src/views/explorer/explorer.module.styl#L6)):*
> ```Stylus
> // included in the /LICENSE file.
> 
> @require '~src/components/core/index.styl'
> ```

sorry to be a pest, can we use scss instead? I want to deprecate stylus.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgXgrtk-taTUF7HQped:-Ogby-CLASUhqX8Zan5Z:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you add a bit more detail? I think it should explain:

1. There's a heatmap where each node is represented by a hexagon.
2. User has ability to select a metric which determines what the colors on the hexagons display
3. Details about what metric is in the histogram. Same as the one in the hexagons?

It's not entirely clear how the selected metric is shown across nodes and also in the histogram across ranges. An explanation of your intent here would be handy.
</blockquote></details>

done
___
*[`-- commits` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-Ogbq1ks1IVJpL9wDd_y:-OgbxzSg9h_mTZey-hqT:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This page should not be documented in a release note. It's not ready for customer consumption.
</blockquote></details>

done
___
*[`pkg/ui/workspaces/db-console/src/views/explorer/explorer.module.styl` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-OgbpGM18DlbUXHnP4dj:-OgbxsCQD04gGIgbXGoX:biuzebq) ([raw file](https://github.com/cockroachdb/cockroach/blob/93de803347c22b056f79653112b9340d02197ca3/pkg/ui/workspaces/db-console/src/views/explorer/explorer.module.styl#L6)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

sorry to be a pest, can we use scss instead? I want to deprecate stylus.
</blockquote></details>

discussed, will leave as stylus due to scss not being currently supported


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
ui: add new cluster explorer page to advanced debug

This new page contains a new hexagon heatmap powered by d3.
Each hexagon represents a single node in the cluster, the heatmap is colored by the user-selected
metric: cpu percentage or write bytes.

When a node hexagon is selected, a uplot histogram is shown.
This histogram offers further drill-down of the same metric on ranges on the selected node.
This histogram component is similar to the existing BarGraphTimeSeries component.

Epic: None
Release note: None

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @dhartunian made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158546#-:-Oh67bPB6rTm1EyeK5yM:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @kyle-a-wong).



<!-- Sent from Reviewable.io -->
